src/mds/MDSRank.cc

```diff
@@ -2905,47 +2907,51 @@ void MDSRankDispatcher::handle_asok_command(
  */
 void MDSRankDispatcher::evict_clients(
   const SessionFilter &filter,
-  std::function<void(int,const std::string&,bufferlist&)> on_finish)
+  std::function<void(int,const std::string&,bufferlist&)> on_finish, bool yes_i_really_mean_it)
 {
   bufferlist outbl;
   if (is_any_replay()) {
     on_finish(-CEPHFS_EAGAIN, "MDS is replaying log", outbl);
     return;
   }
 
-  std::vector<Session*> victims;
-  const auto& sessions = sessionmap.get_sessions();
-  for (const auto& p : sessions)  {
-    if (!p.first.is_client()) {
-      continue;
-    }
+  if (yes_i_really_mean_it) {
+    std::vector<Session*> victims;
+    const auto& sessions = sessionmap.get_sessions();
+    for (const auto& p : sessions)  {
+      if (!p.first.is_client()) {
+        continue;
+      }
 
-    Session *s = p.second;
+      Session *s = p.second;
 
-    if (filter.match(*s, std::bind(&Server::waiting_for_reconnect, server,
-				   std::placeholders::_1))) {
-      victims.push_back(s);
+      if (filter.match(*s, std::bind(&Server::waiting_for_reconnect, server,
+            std::placeholders::_1))) {
+        victims.push_back(s);
+      }
     }
-  }
 
-  dout(20) << __func__ << " matched " << victims.size() << " sessions" << dendl;
+    dout(20) << __func__ << " matched " << victims.size() << " sessions" << dendl;
 
-  if (victims.empty()) {
-    on_finish(0, {}, outbl);
-    return;
-  }
+    if (victims.empty()) {
+      on_finish(0, {}, outbl);
+      return;
+    }
 
-  C_GatherBuilder gather(g_ceph_context,
-			 new LambdaContext([on_finish](int r) {
-					     bufferlist bl;
-					     on_finish(r, {}, bl);
-					   }));
-  for (const auto s : victims) {
-    CachedStackStringStream css;
-    evict_client(s->get_client().v, false,
-                 g_conf()->mds_session_blocklist_on_evict, *css, gather.new_sub());
+    C_GatherBuilder gather(g_ceph_context,
+        new LambdaContext([on_finish](int r) {
+                bufferlist bl;
+                on_finish(r, {}, bl);
+              }));
+    for (const auto s : victims) {
+      CachedStackStringStream css;
+      evict_client(s->get_client().v, false,
+                  g_conf()->mds_session_blocklist_on_evict, *css, gather.new_sub());
+    }
+    gather.activate();
   }
-  gather.activate();
+  dout(20) << __func__ << " doesn't have 'yes_i_really_mean_it' flag passed " << dendl;
+  return;
```

Comment:
I think a better approach would be to return as soon we find that `yes_i_really_mean_it` is missing; for this I'd suggest making `MDSRankDispatcher::evict_clients` return an int val, and have something like:

```
r = evict_clients(filter, on_finish, yes_i_really_mean_it);
if (r < 0) {
   *css << "evicting clients require yes_i_really_mean_it to be passed";
   goto out;
}
```

This would help understand user on why it didn't evict client and prevents digging into logs.

Reply:
True!! Will update. thanks